cobalt/renderer/cobalt_render_frame_observer.cc

```diff
@@ -0,0 +1,138 @@
+// found in the LICENSE file.
+
+#include "cobalt/renderer/cobalt_render_frame_observer.h"
+
+#include "cobalt/configuration/configuration.h"
+#include "content/public/common/isolated_world_ids.h"
+#include "content/public/renderer/chrome_object_extensions_utils.h"
+#include "content/public/renderer/render_frame.h"
+#include "gin/converter.h"
+#include "starboard/system.h"
+#include "third_party/blink/public/web/blink.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "v8/include/v8.h"
+
+namespace cobalt {
+
+namespace {
+
+enum UserOnExitStrategy {
+  kUserOnExitStrategyClose,
+  kUserOnExitStrategyMinimize,
+  kUserOnExitStrategyNoExit,
+};
+
+uint32_t GetUserOnExitStrategy() {
+  // Convert from the Cobalt gyp setting variable's enum options to the H5VCC
+  // interface enum options.
+  std::string exit_strategy_str(
+      configuration::Configuration::GetInstance()->CobaltUserOnExitStrategy());
+#if SB_IS(EVERGREEN)
+  // Note: The status string used here must be synced with the
+  // ComponentState::kUpdated status string defined in updater_module.cc.
+  if (exit_strategy_str == "suspend" &&
+      updater_->GetUpdateStatus() == "Update installed, pending restart") {
+    return static_cast<UserOnExitStrategy>(kUserOnExitStrategyClose);
+  }
+#endif
+  if (exit_strategy_str == "suspend") {
+    return static_cast<UserOnExitStrategy>(kUserOnExitStrategyMinimize);
+  }
+  if (exit_strategy_str == "noexit") {
+    return static_cast<UserOnExitStrategy>(kUserOnExitStrategyNoExit);
+  }
+  if (exit_strategy_str != "stop") {
+    NOTREACHED() << "Unknown exit strategy.";
+  }
+  return static_cast<UserOnExitStrategy>(kUserOnExitStrategyClose);
+}
+
+void PerformUserOnExitStrategy() {
+  if (GetUserOnExitStrategy() == kUserOnExitStrategyMinimize) {
+    SbSystemRequestConceal();
+    return;
+  }
+
+  if (GetUserOnExitStrategy() == kUserOnExitStrategyClose) {
+    SbSystemRequestStop(0);
+    return;
+  }
+}
+
+}  // namespace
+
+CobaltRenderFrameObserver::CobaltRenderFrameObserver(
+    content::RenderFrame* render_frame,
+    base::RepeatingClosure user_on_exit_strategy_callback)
+    : content::RenderFrameObserver(render_frame) {
+  user_on_exit_strategy_callback_ =
+      user_on_exit_strategy_callback
+          ? user_on_exit_strategy_callback
+          : base::BindRepeating(&PerformUserOnExitStrategy);
+}
+
+void CobaltRenderFrameObserver::DidCreateScriptContext(
+    v8::Local<v8::Context> v8_context,
+    int32_t world_id) {
+  // Only install on the main world context of the main frame.
+  if (!render_frame() || !render_frame()->IsMainFrame() ||
+      world_id != content::ISOLATED_WORLD_ID_GLOBAL) {
+    return;
+  }
+
+  InstallCloseAndMinimize();
+}
+
+// static
+void CobaltRenderFrameObserver::DoWindowCloseOrMinimize(
+    const v8::FunctionCallbackInfo<v8::Value>& info) {
+  CobaltRenderFrameObserver* self = static_cast<CobaltRenderFrameObserver*>(
+      v8::External::Cast(*info.Data())->Value());
```

Comment:
nit: imo `info->Data()` is slightly more readable/debuggable than `*info.Data()`

Reply:
Ah right, my mistake